File: ros/hbba_lite/src/hbba_lite/filter_state.py
from abc import ABC, abstractmethod

import rospy

from hbba_lite.srv import SetOnOffFilterState, SetOnOffFilterStateResponse, SetThrottlingFilterState, SetThrottlingFilterStateResponse


class _HbbaFilterState(ABC):
    @abstractmethod
    def check(self):
        return False


class OnOffHbbaFilterState(_HbbaFilterState):
    def __init__(self, state_service_name):
        self._state_service = rospy.Service(state_service_name, SetOnOffFilterState, self._state_service_callback)
        self._is_filtering_all_messages = True

    def _state_service_callback(self, request):
        self._is_filtering_all_messages = request.is_filtering_all_messages
        return SetOnOffFilterStateResponse(ok=True)

    def check(self):
        return not self._is_filtering_all_messages


class ThrottlingHbbaFilterState(_HbbaFilterState):
    def __init__(self, state_service_name):
        self._state_service = rospy.Service(state_service_name, SetThrottlingFilterState, self._state_service_callback)
        self._is_filtering_all_messages = True
        self._rate = 1
        self._counter = 0

    def _state_service_callback(self, request):
        if request.rate <= 0 :
            return SetThrottlingFilterStateResponse(ok=False)

        self._is_filtering_all_messages = request.is_filtering_all_messages
        self._rate = request.rate
        self._counter = 0
        return SetThrottlingFilterStateResponse(ok=True)

    def check(self):
        if self._is_filtering_all_messages:
            return False

        is_ready = False
        if self._counter == 0:
            is_ready = True
        self._counter = (self._counter + 1) % self._rate

        return is_ready

File: ros/hbba_lite/src/hbba_lite_cpp/FilterState.cpp
#include <hbba_lite/FilterState.h>

using namespace std;

OnOffHbbaFilterState::OnOffHbbaFilterState(ros::NodeHandle& nodeHandle, const std::string& stateServiceName) :
    m_isFilteringAllMessages(true)
{
    m_stateService = nodeHandle.advertiseService(stateServiceName, &OnOffHbbaFilterState::stateServiceCallback, this);
}

bool OnOffHbbaFilterState::check()
{
    return !m_isFilteringAllMessages;
}

bool OnOffHbbaFilterState::stateServiceCallback(hbba_lite::SetOnOffFilterState::Request &request,
        hbba_lite::SetOnOffFilterState::Response &response)
{
    m_isFilteringAllMessages = request.is_filtering_all_messages;
    response.ok = true;
    return true;
}

ThrottlingHbbaFilterState::ThrottlingHbbaFilterState(ros::NodeHandle& nodeHandle, const std::string& stateServiceName) :
    m_isFilteringAllMessages(true), m_rate(1), m_counter(0)
{
    m_stateService = nodeHandle.advertiseService(stateServiceName, &ThrottlingHbbaFilterState::stateServiceCallback, this);
}

bool ThrottlingHbbaFilterState::check()
{
    if (m_isFilteringAllMessages)
    {
        return false;
    }

    bool isReady = false;
    if (m_counter == 0)
    {
        isReady = true;
    }
    m_counter = (m_counter + 1) % m_rate;
    return isReady;
}

bool ThrottlingHbbaFilterState::stateServiceCallback(hbba_lite::SetThrottlingFilterState::Request &request,
        hbba_lite::SetThrottlingFilterState::Response &response)
{
    if (request.rate <= 0)
    {
        response.ok = false;
    }
    else
    {
        m_isFilteringAllMessages = request.is_filtering_all_messages;
        m_rate = request.rate;
        m_counter = 0;

        response.ok = true;
    }

    return true;
}
